Allow injection of Gson or JsonParser (or both) when constructing JsonHubProtocol

hcshmk opened the pull request:
# Allow injection of Gson or JsonParser (or both) when constructing JsonHubProtocol

This PR allows the user to inject a `Gson` or `JsonParser` (or both) instance when constructing a `JsonHubProtocol`.

## Why
It is currently not possible (or at least very difficult) to transfer model objects with fields of non-primitive types over SignalR, such as `LocalDateTime`. Allowing the user to add more parsing options to the `Gson` instance before constructing the `JsonHubProtocol` solves this issue.

## Compatibility 
This change is non-breaking and modifies very few lines of code. This change adds additional constructors to `JsonHubProtocol`, and keeps the default constructor to ensure that there are no behavior changes for all existing usages of the class.

## Changes
### Before
```java
public JsonHubProtocol() {
    this.gson = new Gson();
    this.jsonParser = new JsonParser();
}
```
### After
```java
public GsonHubProtocol(Gson gson) {
    this(gson, new JsonParser());
}
```

## References
Addresses https://github.com/dotnet/aspnetcore/issues/17519

Related to https://github.com/dotnet/aspnetcore/pull/17774

Commit:
Allow injection of Gson or JsonParser (or both) when constructing JsonHubProtocol

davidfowl commented:
Can you add a test for this?

halter73 commented:
+1 for tests. One thing to note is that `JsonHubProtocol` is currently [package-private](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html) unlike `MessagePackHubProtocol` which is public. The tests have access to package-private classes, so you shouldn't need to change the accessibility for the tests to work, but it will be important to make this API usable outside of tests.

I filled out the [API proposal template](https://github.com/dotnet/aspnetcore/issues/new?assignees=&labels=api-suggestion&template=30_api_proposal.md&title=) in the issue [here](https://github.com/dotnet/aspnetcore/issues/17519#issuecomment-1347425469). I'd like to get your feedback. I think we can get away with just adding the two-argument `JsonHubProtocol(Gson gson, JsonParser jsonParser)` constructor instead of all four. I also think it might be a good idea to rename `JsonHubProtocol` to `GsonHubProtocol` if we're taking Gson-specific constructor parameters.

hcshmk commented:
@halter73
> One thing to note is that `JsonHubProtocol` is currently [package-private](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html) unlike `MessagePackHubProtocol` which is public.

I'll change the visibility to public.

> I also think it might be a good idea to rename `JsonHubProtocol` to `GsonHubProtocol` if we're taking Gson-specific constructor parameters.

I specifically avoided renaming anything to avoid breaking existing usages of the class, but I do agree with you that it's tempting. I'm renaming it to `GsonHubProtocol` now and will probably follow up with a new `JacksonHubProtocol` later since Jackson is _much_ nicer to work with.

A note on naming - This implementation of `HubProtocol` talks/transmits/interprets JSON data, not Gson data (no such thing) so I'd argue that the rename is misleading. SignalR is known to talk messagepack and json, but to actually get the java client to talk JSON you'd have to know that the implementation uses gson for serialization. It's just a thought.
1. Breaks existing usages
2. Slightly misleading / confusing

> I think we can get away with just adding the two-argument `JsonHubProtocol(Gson gson, JsonParser jsonParser)` constructor instead of all four.

The constuctor `JsonHubProtocol(JsonParser jsonParser)` is probably overkill, especially since you're not even supposed to use instances of `JsonParser`. I'd like to keep `JsonHubProtocol(Gson gson)` since that's the usecase I'm trying to cater to, but you guys have final say on that of course.

@davidfowl
> Can you add a test for this?

Absolutely! What would you like me to test? It seems redundant to test that a constructor without logic constructs the object properly. Do you want me to implement a test showcasing parsing of datetime strings as `LocalDateTime`?

Commit:
Drop unnecessary constructor (Suggested by @halter73)

Commit:
Drop unused import

Commit:
Promote visibility of JsonHubProtocol from package-private to public (Suggested by @halter73)

Commit:
Rename JsonHubProtocol to GsonHubProtocol (Suggested by @halter73)

Commit:
Ensure clean GsonHubProtocol instance for each test

Commit:
Add tests to GsonHubProtocolTests demonstrating deserialization of non-trivial classes using Gson TypeAdaptor (Suggested by @davidfowl)

hcshmk commented:
@halter73 I've implemented the changes you asked for. There was a desire for tests:tm:, but no description of what and why, so I've added tests parsing a regular ISO-8601 datetime string as a `LocalDateTime` object.

hcshmk commented:
@halter73 any news? I've amended my last commit to make [msftbot](https://github.com/apps/msftbot) happy again. It tagged this PR as https://github.com/dotnet/aspnetcore/labels/pending-ci-rerun

halter73 commented:
> @halter73 any news? I've amended my last commit to make [msftbot](https://github.com/apps/msftbot) happy again. It tagged this PR as [ pending-ci-rerun](https://github.com/dotnet/aspnetcore/issues?q=is%3Aopen+label%3Apending-ci-rerun+sort%3Aupdated-desc) When assigned to a PR indicates that the CI checks should be rerun

Yes! Thanks for your patience. The API has been approved, and I commented on the issue with the approved API shape.

We want to make the GsonHubProtocol class `final`. We also think we can do without the `JsonParser` constructor parameter, but we'd be interested in an explanation of the use case for customizing it. We could add it back if it's compelling enough.

BrennanConroy commented:
I checked and JsonParser is actually deprecated and we should be using the static methods instead.
https://www.javadoc.io/doc/com.google.code.gson/gson/latest/com.google.gson/com/google/gson/JsonParser.html
So we definitely don't want it in the public API :smile:

Commit:
Remove GsonHubProtocol constructor taking JsonParser since usage is deprecated in newer versions of gson

hcshmk commented:
> I checked and JsonParser is actually deprecated and we should be using the static methods instead. So we definitely don't want it in the public API

I've removed the constructor that takes `JsonParser` as argument in commit c8329bfe7c8cb9cbf6f013ba779e6bd2b905acba. If you want, I could follow up this PR with one that changes the usage from instance of JsonParser to static functions instead?

In the meantime I'm working on a new `JacksonHubProtocol` based on [Jackson](https://github.com/FasterXML/jackson) instead of Gson, since it's both faster and easier to work with. I'll support my claims with benchmarks.

Commit:
Mark GsonHubProtocol final

hcshmk commented:
The tests runner for Helix 64 fails after almost two hours due to something completely irellevant. Unless you guys know what's wrong, I think I'll just run a new pipeline tomorrow in hope that it's fixed.

![image](https://user-images.githubusercontent.com/69573050/217245428-77d7c9fb-98fa-4baa-a485-652559e35e7b.png)

BrennanConroy commented on src/SignalR/clients/java/signalr/test/src/main/java/com/microsoft/signalr/GsonHubProtocolTest.java:
```diff
@@ -409,7 +417,52 @@ public void errorWhileParsingIncompleteMessage() {
+
+        // Setup appropriate type adaptor
+        Gson gson = new GsonBuilder()
+                .registerTypeAdapter(LocalDateTime.class, ((JsonDeserializer<LocalDateTime>) (json, type, context)
```
Can Gson add custom `int` and `string` adapters? If so, we should add a test that `type`, `target`, and `invocationId` fields aren't affected by Gson config.

hcshmk replied:
You can register type adapters for literally any type, including int and string. The type adapters, however, do not affect the SignalR json protocol stuff, only its payloads.

I've added a test (1937b5f2002b84add8c36f750e36a98d305d031e) with a pretty rude type adapter that attempts to convert any `String` to an `int` using the pattern `"one"` -> `1`, `"two"` -> `2`, `"three"` -> `3`,  default throw exception.

`type`, `target`, and `invocationId` are completely unaffected by this.

BrennanConroy replied:
Does that only test the `type` field? Since the adapter will only run when trying to deserialize Integer and `target` and `invocationId` are strings.

BrennanConroy replied:
> The implementation of `parseMessages` is using manual json parsing (instead of mapping to POJOs) which is why the type adaptor stuff doesnt affect it

Yes, I'm very aware...
That doesn't mean we shouldn't have tests though, so it doesn't accidentally break in the future.

hcshmk replied:
Both `target` and `invocationId` are covered as of b3f97f3c40bbb974ee4e636b14d8366527957c9b. I've added yet another type adapter to the same test, such that ints are converted to strings, and strings are also converted to ints. (The same element won't be converted back and forth, don't worry)
This is a toy example, but I think it covers your concern.

Commit:
Add test confirming that (fairly rude) type adapters do not break signalr json protocol stuff (only affects payload)

Commit:
Add equally bad type adaptor to check that invocationId and target are not mistakenly converted along with payload

BrennanConroy reviewed (approved):
Awesome! Thank you!
